File: Compare_Version_Numbers.cpp
/*
Compare two version numbers version1 and version2.
If version1 > version2 return 1, if version1 < version2 return -1, otherwise return 0.

You may assume that the version strings are non-empty and contain only digits and the . character.
The . character does not represent a decimal point and is used to separate number sequences.
For instance, 2.5 is not "two and a half" or "half way to version three", it is the fifth second-level revision of the second first-level revision.

Here is an example of version numbers ordering:

0.1 < 1.1 < 1.2 < 13.37
*/

class Solution {
public:

    int compareVersion(string version1, string version2) {
       // int size1 = version1.size(), size2=version2.size();
        string subVersion1, subVersion2;
        int inumber1, inumber2, istart1=0, istart2=0, ipos1=0, ipos2=0;
        
        while(istart1 < version1.size() || istart2 < version2.size()){
            if(istart1 >= version1.size())  inumber1 = 0;
            else{
                ipos1 = version1.find_first_of('.', istart1);  //find the first position of '.'
                if(-1 == ipos1) ipos1 = version1.size();  //judge if it is the end of the string
                subVersion1.assign(version1, istart1, ipos1-istart1);  //assign is the function using "." to use
                inumber1 = stoi(subVersion1);  //transfer string to int
            }
            
            if(istart2 >= version2.size())  inumber2 = 0;
            else{
                ipos2 = version2.find_first_of('.', istart2);
                if(-1 == ipos2) ipos2 = version2.size();
                subVersion2.assign(version2, istart2, ipos2-istart2);
                inumber2 = stoi(subVersion2);
            }
            
            if(inumber1 > inumber2) return 1;
            else if(inumber1 < inumber2) return -1;
            
            istart1 = ipos1+1;  //compare the number after '.'
            istart2 = ipos2+1;
            
        }
        return 0;
    }
};

File: Compare_Version_Numbers.swift
/*
Given two version strings, version1 and version2, compare them. A version string consists of revisions separated by dots '.'. The value of the revision is its integer conversion ignoring leading zeros.

To compare version strings, compare their revision values in left-to-right order. If one of the version strings has fewer revisions, treat the missing revision values as 0.

Return the following:

If version1 < version2, return -1.
If version1 > version2, return 1.
Otherwise, return 0.


Example 1:

Input: version1 = "1.2", version2 = "1.10"

Output: -1

Explanation:

version1's second revision is "2" and version2's second revision is "10": 2 < 10, so version1 < version2.

Example 2:

Input: version1 = "1.01", version2 = "1.001"

Output: 0

Explanation:

Ignoring leading zeroes, both "01" and "001" represent the same integer "1".

Example 3:

Input: version1 = "1.0", version2 = "1.0.0.0"

Output: 0

Explanation:

version1 has less revisions, which means every missing revision are treated as "0".



Constraints:

1 <= version1.length, version2.length <= 500
version1 and version2 only contain digits and '.'.
version1 and version2 are valid version numbers.
All the given revisions in version1 and version2 can be stored in a 32-bit integer.
*/


/*
Solution 1:
First get the digit array, then compare it one by one

Time Complexity: O(n1 + n2)
Space Complexity: O(n1 + n2)
- n1 = version1.count
- n2 = version2.count
*/
class Solution {
    func compareVersion(_ version1: String, _ version2: String) -> Int {
        var digits1 = getDigit(version1)
        var digits2 = getDigit(version2)
        let n1 = digits1.count
        let n2 = digits2.count
        var index = 0
        while index < n1 || index < n2 {
            let d1 = index < n1 ? digits1[index] : 0
            let d2 = index < n2 ? digits2[index] : 0
            if (d1 < d2) {
                return -1
            } else if (d1 > d2) {
                return 1
            } else {
                index += 1
            }
        }
        return 0
    }

    // return digit array
    // ex: 1.2 return [1,2]
    func getDigit(_ version: String) -> [Int] {
        var digits = version.split(separator: ".")
        return digits.map { Int($0) ?? 0 }
    }
}
